controllers/buildpipeline/buildpipeline_controller.go

```diff
@@ -0,0 +1,147 @@
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=applications/finalizers,verbs=update
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=applications,verbs=get;list;watch
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=applications/status,verbs=get
+//+kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch
+//+kubebuilder:rbac:groups=pipelinesascode.tekton.dev,resources=repositories,verbs=get;list;watch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := helpers.IntegrationLogger{Logger: r.Log.WithValues("buildpipelineRun", req.NamespacedName)}
+	loader := loader.NewLoader()
+
+	pipelineRun := &tektonv1beta1.PipelineRun{}
+	err := r.Get(ctx, req.NamespacedName, pipelineRun)
+	if err != nil {
+		logger.Error(err, "Failed to get build pipelineRun for", "req", req.NamespacedName)
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{}, err
+	}
+
+	component, err := loader.GetComponentFromPipelineRun(r.Client, ctx, pipelineRun)
+	if err != nil {
+		logger.Error(err, "Failed to get Component for",
+			"PipelineRun.Name", pipelineRun.Name, "PipelineRun.Namespace", pipelineRun.Namespace)
+		return ctrl.Result{}, err
+	}
+
+	application := &applicationapiv1alpha1.Application{}
+	if component != nil {
+		application, err = loader.GetApplicationFromComponent(r.Client, ctx, component)
+		if err != nil {
+			logger.Error(err, "Failed to get Application from Component",
+				"Component.Name ", component.Name, "Component.Namespace ", component.Namespace)
+			return ctrl.Result{}, err
+		}
+	}
+
+	if application == nil {
+		err := fmt.Errorf("failed to get Application")
+		logger.Error(err, "reconcile cannot resolve application")
+		return ctrl.Result{}, err
+	}
+	logger = logger.WithApp(*application)
+
+	adapter := NewAdapter(pipelineRun, component, application, logger, loader, r.Client, ctx)
+
+	return controller.ReconcileHandler([]controller.Operation{
+		adapter.EnsureSnapshotExists,
+	})
+}
+
+// AdapterInterface is an interface defining all the operations that should be defined in an Integration adapter.
+type AdapterInterface interface {
+	EnsureSnapshotExists() (controller.OperationResult, error)
+}
+
+// SetupController creates a new Integration reconciler and adds it to the Manager.
+func SetupController(manager ctrl.Manager, log *logr.Logger) error {
+	return setupControllerWithManager(manager, NewIntegrationReconciler(manager.GetClient(), log, manager.GetScheme()))
+}
+
+// setupCache indexes fields for each of the resources used in the pipeline adapter in those cases where filtering by
+// field is required.
+func setupCache(mgr ctrl.Manager) error {
+	if err := cache.SetupApplicationComponentCache(mgr); err != nil {
```

Comment:
Since we already set up the same cache in the integrationpipeline controller [here](Image repository where the built image was pushed) (temporary link), this causes a conflict.
Basically, we cannot set up the same type of cache in multiple places.

Reply:
Thanks! Let me update it

Reply:
Even after pushing this change, I saw the CI fail with that same bug, am I right?
If yes, what was/were the other breaking change(s)?

Reply:
Both of the two cache functions should be removed.